swirl_lm/boundary_condition/immersed_boundary_method.py

```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
+      # fluid domain, and `0` if inside the solid domain.
+      # Source: Iaccarino and Verzicco, Applied Mechanics Reviews 56(3), 2003.
+      degenerated_cell_mask = tf.nest.map_structure(
+          lambda v: tf.where(v < 7.5, tf.ones_like(v), tf.zeros_like(v)),
+          tf.math.reduce_max(fluid_corner_point_mask, axis=0)
+      )
+
+      # Add the boundary interecpt (BI) as an additional point, if the enclosing
+      # cell of a given image point is degenerated (i.e. has less than 8
+      # enclosing nodes in the fluid domain).
+      # Essentially, we are adding the distance between the ghost point(!) and
+      # the IB to `node_dist` which will now have 9 rows (8 distances between the
+      # image point and its surrounding grid nodes, and 1 additional row for the
+      # distance between the image point and the IB). However, the 9th row will
+      # only have non-zero distances, if the enclosing grid cell is degenerated,
+      # otherwise, the distance will be zero. As we will always use all 9 rows of
+      # `node_dist` to compute the interpolated value on the image point, we want
+      # to make sure that the BI is excluded for an eclosing cell that has all 8
+      # grid nodes in the fluid domain, and included otherwise.
+      node_dist.append(
+          tf.nest.map_structure(
+            lambda d, m: tf.math.pow(
+                tf.math.divide(1.0, d + _EPSILON), _EXP) * m,
+            gp_dist,
+            degenerated_cell_mask
+          )
+      )
+
+      # Get the maximum distance between the image point and its enclosing
+      # grid nodes. We only consider grid nodes in the fluid domain, and
+      # discard nodes in the solid domain.
+      node_dist_max = tf.math.reduce_max(node_dist, axis=0)
+
+      # Next, we compute the corresponding interpolation weight of each enclosing
+      # grid node as follows:
+      # gamma_i = ((node_dist_max_i - node_dist_i) / (
+      #     node_dist_max_i * node_dist_i)
+      # )**0.5
+      # where gamma_i is the un-normalized interpolation weight of the i-th
+      # enclosing grid node for a given image point. The normalized weight is
+      # then obtained by
+      # w = (1 / denom) * ∑_k gamma_k.
+      # Note that we always exclude the body intercept weight for Neumann boundary
+      # conditions. This is because the value AT the body intercept is not known
+      # before executing the interpolation of the enclosing grid nodes onto the
+      # image point. Consequently, we need a separate set of interpolation weights
+      # for Neumann boundary conditions.
+      # Reference: R. Franke, Scattered data interpolation: tests of some
+      # methods, Mathematics of Computation 38(157), 1982,
+      # https://www.ams.org/mcom/1982-38-157/S0025-5718-1982-0637296-4/
+      # And:
+      # Cai et al., JCP 429, 2021, specifically Eq. 35, p. 9
+
+      # For Dirichlet boundary conditions:
+      gamma = tf.nest.map_structure(
```

Comment:
Let's make the gamma computation a helper function at the file level so that we can test it with artificial inputs.